File: 3532-path-existence-queries-in-a-graph-i/3532-path-existence-queries-in-a-graph-i.cpp
class Solution {
public:
    vector<bool> pathExistenceQueries(int n, vector<int>& nums, int maxDiff,
                                      vector<vector<int>>& queries) {
        vector<int> group(n, 0);
        int currentGroup = 0;
        for (int i = 1; i < n; ++i) {
            if (nums[i] - nums[i - 1] > maxDiff) {
                currentGroup++;
            }
            group[i] = currentGroup;
        }

        vector<bool> res;
        for (auto& q : queries) {
            int u = q[0], v = q[1];
            res.push_back(group[u] == group[v]);
        }
        return res;
    }
};

File: 3532-path-existence-queries-in-a-graph-i/3532-path-existence-queries-in-a-graph-i.java
class Solution {
        public boolean[] pathExistenceQueries(int n, int[] A, int maxDiff, int[][] queries) {
        int[] g = new int[n];
        for (int i = 1; i < n; i++) {
            g[i] = g[i - 1] + ((A[i] - A[i - 1]) > maxDiff ? 1 : 0);
        }
        int qn = queries.length, qi = 0;
        boolean[] res = new boolean[qn]; 
        for (int[] q: queries) {
            res[qi++] = g[q[0]] == g[q[1]];
        }
        return res;
    }
}